### PR TITLE
FILTER: Crop Image Geometry

### PR DESCRIPTION
**Pull request by mmarineBlueQuartz**
* Added CropImageGeometry filter
* Added SamplingUtils
* Added unit test

### CODE REVIEW

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/CropImageGeometry.cpp [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+  }
+  if(zMin < 0)
+  {
+    std::string ss = fmt::format("Z Min ({}) less than 0", zMin);
+    return {MakeErrorResult<OutputActions>(-5550, ss)};
+  }
+
+  // Validate the incoming DataContainer, Geometry, and AttributeMatrix ; bail if any do not exist since we plan on using them later on in the dataCheck
+  auto oldDimensions = getCurrentVolumeDataContainerDimensions(data, srcImagePath);
+  auto oldResolution = getCurrentVolumeDataContainerResolutions(data, srcImagePath);
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/CropImageGeometry.cpp:233]: The comment indicates that these are required, but the functions don't have a way to report errors.

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/CropImageGeometry.cpp [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+    auto action = std::make_unique<CreateImageGeometryAction>(destImagePath, tDims, newOrigin, spacingVec);
+    actions.actions.push_back(std::move(action));
+
+    auto newCellFeaturesPath = destImagePath.createChildPath(newCellFeaturesName);
+    {
+      auto action = std::make_unique<CreateDataGroupAction>(newCellFeaturesPath);
+      actions.actions.push_back(std::move(action));
+    }
+
+    auto* srcImageGeom = data.getDataAs<ImageGeom>(srcImagePath);
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/CropImageGeometry.cpp:304]: This line shadows the variable on line 235 and looks like it's getting the same geometry anyway.

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/CropImageGeometry.cpp [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+      auto action = std::make_unique<CreateArrayAction>(numericType, std::vector<usize>{numTuples}, std::vector<usize>{components}, dataArrayPath);
+      actions.actions.push_back(std::move(action));
+    }
+  }
+
+  usize totalPoints = srcImageGeom->getNumberOfElements();
+
+  if(shouldRenumberFeatures)
+  {
+    std::vector<usize> cDims = {1};
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/CropImageGeometry.cpp:327]: Is this meant to be used to check the number of components for the following array?

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/CropImageGeometry.cpp [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+  auto zMax = args.value<int32>(k_MaxZ_Key);
+  auto zMin = args.value<int32>(k_MinZ_Key);
+  auto shouldUpdateOrigin = args.value<bool>(k_UpdateOrigin_Key);
+  auto shouldRenumberFeatures = args.value<bool>(k_RenumberFeatures_Key);
+  auto newFeaturesName = args.value<std::string>(k_NewFeaturesName_Key);
+  auto voxelArrayPaths = args.value<std::vector<DataPath>>(k_VoxelArrays_Key);
+  auto featureIdsArrayPath = args.value<DataPath>(k_FeatureIds_Key);
+
+  auto* srcImageGeom = data.getDataAs<ImageGeom>(srcImagePath);
+  auto* destImageGeom = data.getDataAs<ImageGeom>(destImagePath);
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/CropImageGeometry.cpp:359]:
```suggestion
  auto& srcImageGeom = data.getDataRefAs<ImageGeom>(srcImagePath);
  auto& destImageGeom = data.getDataRefAs<ImageGeom>(destImagePath);
```

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

**Review by imikejackson — CHANGES_REQUESTED** [review]
Not much more to say past what @JDuffeyBQ has stated. Ensure the unit test for this works or we have bench marked this against the current version in SIMPL/DREAM.3D

The "RenumberFeatures" should also be generic enough that other filters can use it. I'm pretty sure that a handful of filters in SIMPL/DREAM.3D need this functionality.

**Review by JDuffeyBQ — COMMENTED** [review]
This will need a minor update for the cancellation changes.

**Review by JDuffeyBQ — APPROVED** [review]

**Review by JDuffeyBQ — DISMISSED** [review]

**Comment by imikejackson** [comment]
Is this ready to go? Not seen any activity. @mmarineBlueQuartz

**Comment by imikejackson** [comment]
Also make sure to bring over the documentation file (plus any references images) from current DREAM3D

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/CropImageGeometry.cpp [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+  FloatVec3 data = {0, 0, 0};
+  const auto* image = dataStructure.getDataAs<ImageGeom>(imageGeomPath);
+  if(image != nullptr)
+  {
+    data = image->getSpacing();
+  }
+  return data;
+}
+
+std::string getOldBoxDimensions(const DataStructure& dataStructure, const DataPath& imageGeomPath)
```
Comment (JDuffeyBQ) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/CropImageGeometry.cpp:109]: These two functions aren't used right now, but in the original the were used for the preflight updated value.

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

**Review by JDuffeyBQ — APPROVED** [review]

**Review by imikejackson — APPROVED** [review]

**Pull request merged**